### PR TITLE
style: Implement finer-grained stylist rebuilds.

### PR DESCRIPTION
**Pull request by emilio**
This is on top of #18170, and aims to resolve https://bugzilla.mozilla.org/show_bug.cgi?id=1386045.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18191)

### CODE REVIEW

**Comment by emilio** [comment]
@bors-servo try

https://treeherder.mozilla.org/#/jobs?repo=try&revision=984dbde106112dfe1378681a01dc9270229e8a1a&selectedJob=124893656

**Comment by emilio** [comment]
@bors-servo try

 * Generic exception, so I ammended without touching anything and pushed again.

**Comment by emilio** [comment]
Rebased.

@bors-servo try

**Comment by emilio** [comment]
@bors-servo try

Rebased.

**Comment by emilio** [comment]
@bors-servo try

 * Generic exception?

File: components/style/stylesheet_set.rs [hunk 1]
```diff
@@ -56,26 +79,34 @@ where
...
-    /// partial rebuilds.
+    /// A full rebuild, of both cascade data and invalidation data.
     Full,
+    /// A partial rebuild, of only the cascade data.
+    CascadeOnly,
+}
+
+impl SheetRebuildKind {
+    /// Whether the stylesheet invalidation data should be rebuilt.
+    pub fn rebuild_invalidation(&self) -> bool {
```
Comment (heycam) [inline components/style/stylesheet_set.rs]: Nit: function names sounds a bit like it's going to do the rebuilding. Maybe "must_rebuild_invalidation" or something?

File: components/style/stylist.rs [hunk 2]
```diff
@@ -98,15 +98,24 @@ impl DocumentCascadeData {
...
+
+            if origin == Origin::UserAgent {
+                self.precomputed_pseudo_element_decls.clear();
+            }
 
-        if origins_to_rebuild.contains(Origin::UserAgent.into()) {
-            self.precomputed_pseudo_element_decls.clear();
+            extra_data.borrow_mut_for_origin(&origin).clear();
+            if validity == OriginValidity::CascadeInvalid {
+                cascade_data.clear_cascade_data()
```
Comment (heycam) [inline components/style/stylist.rs]: Hmm, now these names are slightly confusing I guess.  ("Cascade data" both for the top level, per-origin data, and also for the tables which represent performing the cascading and not the invalidations...)

File: components/style/stylist.rs [hunk 3]
```diff
@@ -119,15 +128,26 @@ impl DocumentCascadeData {
...
-                    self.add_stylesheet(
-                        device,
-                        quirks_mode,
-                        stylesheet,
-                        guards.ua_or_user,
-                        extra_data,
-                    );
+                let validity = flusher.origin_validity(sheet_origin);
+
+                // Servo doesn't support to incrementally mutate UA sheets.
```
Comment (heycam) [inline components/style/stylist.rs]: To reinforce this, would it make sense to assert in here that if ua_stylesheets is non-None that this is not stylo?

File: components/style/stylist.rs [hunk 4]
```diff
@@ -140,25 +160,35 @@ impl DocumentCascadeData {
                     Origin::UserAgent | Origin::User
                 ));
 
-                if origins_to_rebuild.contains(sheet_origin.into()) {
+                let validity = flusher.origin_validity(sheet_origin);
+
+                // Servo doesn't support to incrementally mutate UA sheets.
```
Comment (heycam) [inline components/style/stylist.rs]: And in here?

**Comment by SimonSapin** [comment]
Nice work!

@bors-servo r+

---

Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 2 files at r6, 2 of 2 files at r8, 1 of 1 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/18191#-:-KsDkHkkPXVHzHBKusgm:br5kgb5)*

**Review by heycam — APPROVED** [review]
Looks good!

File: components/style/stylesheet_set.rs [hunk 5]
```diff
@@ -9,16 +9,26 @@ use invalidation::stylesheets::StylesheetInvalidationSet;
 use media_queries::Device;
 use shared_lock::SharedRwLockReadGuard;
 use std::slice;
-use stylesheets::{OriginSet, PerOrigin, StylesheetInDocument};
+use stylesheets::{Origin, OriginSet, PerOrigin, StylesheetInDocument};
 
 /// Entry for a StylesheetSet. We don't bother creating a constructor, because
 /// there's no sensible defaults for the member variables.
```
Comment (waywardmonkeys) [inline components/style/stylesheet_set.rs]: I think this comment is out of date now that there's a `new` method for this struct?

File: components/style/stylist.rs (hunk 2 above)
Comment (emilio) [inline components/style/stylist.rs]: Yeah, I don't have any better naming idea though... I can rename the current `CascadeData` to `StyleData`, and divide it in `CascadeData` and `InvalidationData` I guess... But that seemed a bit overkill, I'll file a followup.

File: components/style/stylesheet_set.rs (hunk 1 above)
Comment (emilio) [inline components/style/stylesheet_set.rs]: It made lines slightly longer than what I felt comfortable with, but sure.

File: components/style/stylesheet_set.rs (hunk 5 above)
Comment (emilio) [inline components/style/stylesheet_set.rs]: Good catch

File: components/style/stylist.rs (hunk 3 above)
Comment (emilio) [inline components/style/stylist.rs]: I'll assert `cfg!(feature = "servo")` above.

**Pull request merged**